Add xla autocast support, update autocast APIs in checkpointing

savitha-aws opened the pull request:
Changes to gradient checkpointing: 

1. Add xla autocast support 
2. Update use of deprecated autocast APIs

rpsilva-aws commented:
FYI @ManfeiBai @tengyifei

rpsilva-aws commented:
@savitha-aws If this resolves autocast with checkpointing, do we have a way to test it? Can we write a short test with the MLP model if possible?

savitha-aws commented:
@rpsilva-aws Modified the existing gradient checkpointing test to include an autocast option. Also added this to neuron tests.

rpsilva-aws commented:
> @rpsilva-aws Modified the existing gradient checkpointing test to include an autocast option. Also added this to neuron tests.

Can we also add it for `tpu/run_tests.sh`, so it also runs as part of the workflows? For other folks, it seems that it isn't there for CPU, because it doesn't work with optimization barriers (likely why it was intentionally left out of the standard run_tests script).

savitha-aws commented:
> > @rpsilva-aws Modified the existing gradient checkpointing test to include an autocast option. Also added this to neuron tests.
> 
> Can we also add it for `tpu/run_tests.sh`, so it also runs as part of the workflows? For other folks, it seems that it isn't there for CPU, because it doesn't work with optimization barriers (likely why it was intentionally left out of the standard run_tests script).

Added to tpu tests as well.

bhavya01 commented:
@savitha-aws The grad checkpointing test is currently failing in our CI with the error:

```
Traceback (most recent call last):
  File "/home/bbahl/xla/test/test_grad_checkpoint.py", line 50, in <module>
    run()
  File "/home/bbahl/xla/test/test_grad_checkpoint.py", line 44, in run
    dummy_loss.backward()
  File "/home/bbahl/miniconda3/envs/torch26/lib/python3.10/site-packages/torch/_tensor.py", line 626, in backward
    torch.autograd.backward(
  File "/home/bbahl/miniconda3/envs/torch26/lib/python3.10/site-packages/torch/autograd/__init__.py", line 347, in backward
    _engine_run_backward(
  File "/home/bbahl/miniconda3/envs/torch26/lib/python3.10/site-packages/torch/autograd/graph.py", line 823, in _engine_run_backward
    return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
  File "/home/bbahl/miniconda3/envs/torch26/lib/python3.10/site-packages/torch/autograd/function.py", line 307, in apply
    return user_fn(self, *args)
  File "/home/bbahl/miniconda3/envs/torch26/lib/python3.10/site-packages/torch_xla/utils/checkpoint.py", line 185, in backward
    outputs = ctx.run_function(*detached_inputs)
  File "/home/bbahl/miniconda3/envs/torch26/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1739, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/bbahl/miniconda3/envs/torch26/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1750, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/bbahl/miniconda3/envs/torch26/lib/python3.10/site-packages/torch/nn/modules/container.py", line 250, in forward
    input = module(input)
  File "/home/bbahl/miniconda3/envs/torch26/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1739, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/bbahl/miniconda3/envs/torch26/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1750, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/bbahl/miniconda3/envs/torch26/lib/python3.10/site-packages/torch/nn/modules/conv.py", line 554, in forward
    return self._conv_forward(input, self.weight, self.bias)
  File "/home/bbahl/miniconda3/envs/torch26/lib/python3.10/site-packages/torch/nn/modules/conv.py", line 549, in _conv_forward
    return F.conv2d(
RuntimeError: Input type (c10::BFloat16) and bias type (float) should be the same
```
My understanding that during the backward pass, we run the forward function again but the input dtype is bf16 where as the bias dtype is float32. Since, it is not run with the `autocast` context, we get the dtype mismatch error. Can you please take a look?

savitha-aws commented:
Hi @bhavya01, 

I'm not able to reproduce this error on a Neuron device. I see both weights and biases being converted to bf16 in the HLO. I also tried the test on TPU, and it seems to pass. Is this an internal CI that the test fails?

bhavya01 commented:
> Hi @bhavya01,
> 
> I'm not able to reproduce this error on a Neuron device. I see both weights and biases being converted to bf16 in the HLO. I also tried the test on TPU, and it seems to pass. Is this an internal CI that the test fails?

Thanks for verifying! We realized that CI tests were running the last built release which was before this change was merged but test source code was taken from r2.6 branch head causing the test to fail. 

It's all sorted now.
